### PR TITLE
Pr ekf rng fuse updates2

### PR DESCRIPTION
**Pull request by tumbili**
@devbharat @priseborough This is a replacement for #216 

@devbharat Check the last commit I did. I got rid of the hrt dependency and the static variables and fixed some other stuff.
I gave it a test in SITL and the result was as expected. Can you have a look?

### CODE REVIEW

**Comment by tumbili** [comment]
@devbharat @priseborough I rebased this on master, can somebody please do a review so that we can bring it in?

File: EKF/terrain_estimator.cpp [hunk 1]
```diff
@@ -69,37 +70,52 @@ bool Ekf::initHagl()
...
 		// predict the hagl from the vehicle position and terrain height
 		float pred_hagl = _terrain_vpos - _state.pos(2);
 
 		// calculate the innovation
 		_hagl_innov = pred_hagl - meas_hagl;
 
-		// calculate the observation variance adding the variance of the vehicles own height uncertainty and factoring in the effect of tilt on measurement error
-		float obs_variance = fmaxf(P[9][9], 0.0f) + sq(_params.range_noise / _R_to_earth(2, 2));
+		// calculate the observation variance adding the variance of the vehicles own height uncertainty
+		float obs_variance = fmaxf(P[9][9], 0.0f) + sq(_params.range_noise) + sq(_params.range_noise_scaler * _range_sample_delayed.rng);
```
Comment (tumbili) [inline EKF/terrain_estimator.cpp]: @priseborough Shouldn't we be multiplying the last term in the upper line with sq(_R_to_earth(2, 2))?

**Comment by LorenzMeier** [comment]
@priseborough @ChristophTobler Can you please review?

File: EKF/terrain_estimator.cpp (hunk 1 above)
Comment (priseborough) [inline EKF/terrain_estimator.cpp]: For a level flat earth without tilt estimation and range finder alignment errors  where the only error source is scale factor n the range finder the answer is yes. However terrain height variation underneath the vehicle and attitude errors do increase the measurement height uncertainty with range.  The derivative from vehicle tilt error and terrain tilt to height scales with range * sin(tilt angle). This tends to cancel the effect of range scale errors being 'compressed' with tilt.

**Comment by tumbili** [comment]
@priseborough I pushed two more commits. I noticed that the sin and cos values of the range pitch angle were only calculated once on initialisation but not after. This means that parameter changes of the range sensor pitch angle were not having any effect.

File: EKF/control.cpp [hunk 2]
```diff
@@ -84,6 +84,10 @@ void Ekf::controlFusionModes()
 	_mag_data_ready = _mag_buffer.pop_first_older_than(_imu_sample_delayed.time_us, &_mag_sample_delayed);
 	_baro_data_ready = _baro_buffer.pop_first_older_than(_imu_sample_delayed.time_us, &_baro_sample_delayed);
 
+	// update range sensor angle parameters in case they have changed
```
Comment (priseborough) [inline EKF/control.cpp]: I would move them into the setRangeData function inside the 

if (time_usec - _time_last_range > _min_obs_interval_us)

conditional so they are not unnecessarily calculated every EKF update.

**Comment by tumbili** [comment]
@priseborough Paul, moving it in the function you proposed does not work since the variables are not members of the base class (EstimatorInterface). I could have moved the variables into the base class but I feel like they don't belong there. I moved the calculations into the terrain estimator where they are executed at the same rate as the fusion happens. What do you think?

**Comment by priseborough** [comment]
Sounds good

**Comment by LorenzMeier** [comment]
@tumbili Can you update your PR and rebase?

**Comment by tumbili** [comment]
@LorenzMeier This was already merged some time ago. Closing.

**Pull request closed without merging**